api/sys_raft.go

```diff
@@ -0,0 +1,158 @@
+package api
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	cleanhttp "github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"golang.org/x/net/http2"
+)
+
+// RaftJoinResponse represents the response of the raft join API
+type RaftJoinResponse struct {
+	Joined bool `json:"joined"`
+}
+
+// RaftJoinRequest represents the parameters consumed by the raft join API
+type RaftJoinRequest struct {
+	LeaderAddr string `json:"leader_api_addr"`
+	CACert     string `json:"ca_cert":`
+	Retry      bool   `json:"retry"`
+}
+
+// RaftJoin adds the node from which this call is invoked from to the raft
+// cluster represented by the leader address in the parameter.
+func (c *Sys) RaftJoin(opts *RaftJoinRequest) (*RaftJoinResponse, error) {
+	r := c.c.NewRequest("POST", "/v1/sys/storage/raft/join")
+
+	if err := r.SetJSONBody(opts); err != nil {
+		return nil, err
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var result RaftJoinResponse
+	err = resp.DecodeJSON(&result)
+	return &result, err
+}
+
+// RaftSnapshot invokes the API that takes the snapshot of the raft cluster and
+// writes it to the supplied io.Writer.
+func (c *Sys) RaftSnapshot(snapWriter io.Writer) error {
+	r := c.c.NewRequest("GET", "/v1/sys/storage/raft/snapshot")
+	r.URL.RawQuery = r.Params.Encode()
+
+	req, err := http.NewRequest(http.MethodGet, r.URL.RequestURI(), nil)
+	if err != nil {
+		return err
+	}
+
+	req.URL.User = r.URL.User
+	req.URL.Scheme = r.URL.Scheme
+	req.URL.Host = r.URL.Host
+	req.Host = r.URL.Host
+
+	if r.Headers != nil {
+		for header, vals := range r.Headers {
+			for _, val := range vals {
+				req.Header.Add(header, val)
+			}
+		}
+	}
+
+	if len(r.ClientToken) != 0 {
+		req.Header.Set(consts.AuthHeaderName, r.ClientToken)
+	}
+
+	if len(r.WrapTTL) != 0 {
+		req.Header.Set("X-Vault-Wrap-TTL", r.WrapTTL)
+	}
+
+	if len(r.MFAHeaderVals) != 0 {
+		for _, mfaHeaderVal := range r.MFAHeaderVals {
+			req.Header.Add("X-Vault-MFA", mfaHeaderVal)
+		}
+	}
+
+	if r.PolicyOverride {
+		req.Header.Set("X-Vault-Policy-Override", "true")
+	}
+
+	config := &Config{
```

Comment:
So since we have access to the actual client here (`c.client`) i wonder if we can reuse that instead of read from the environment a second time? Also in you're using the CLI TLS info may not be coming from the ENV. `c.client.config.Do()` may be all that's needed? Worth testing out

Reply:
Good observation again. Fixed!